In [ ]:
import kagglehub
from google.colab import userdata
import os

# 1. Authorize using the NAMES of your secrets from the sidebar
# We use 'KAGGLE_USERNAME' because that's what we named the secret
os.environ["KAGGLE_USERNAME"] = userdata.get('KAGGLE_USERNAME')
os.environ["KAGGLE_KEY"] = userdata.get('KAGGLE_KEY')

# 2. Download latest version of the Logistics Dataset
path = kagglehub.dataset_download("kundanbedmutha/delivery-logistics-dataset-india-multi-partner")

print("Data is now at:", path)

# 3. List the files in that path so we know what to open
print("\nFiles in folder:", os.listdir(path))



Using Colab cache for faster access to the 'delivery-logistics-dataset-india-multi-partner' dataset.
Data is now at: /kaggle/input/delivery-logistics-dataset-india-multi-partner

Files in folder: ['Delivery_Logistics.csv']


In [ ]:
!pip install --upgrade kagglehub

In [ ]:
import pandas as pd
import os

# We use the path from your previous output
path = "/root/.cache/kagglehub/datasets/kundanbedmutha/delivery-logistics-dataset-india-multi-partner/versions/1"
file_path = os.path.join(path, "Delivery_Logistics.csv")

# Load the data
df = pd.read_csv(file_path)

# 1. See the size of the business (How many rows/columns?)
print(f"Dataset Loaded: {df.shape[0]} delivery records found.\n")

# 2. See the column names (The 'Headers' of our business)
print("Business Columns:", df.columns.tolist())

# 3. View the first 5 records
df.head()



Dataset Loaded: 25000 delivery records found.

Business Columns: ['delivery_id', 'delivery_partner', 'package_type', 'vehicle_type', 'delivery_mode', 'region', 'weather_condition', 'distance_km', 'package_weight_kg', 'delivery_time_hours', 'expected_time_hours', 'delayed', 'delivery_status', 'delivery_rating', 'delivery_cost']


,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,expected_time_hours,delayed,delivery_status,delivery_rating,delivery_cost
0,250.99,delhivery,automobile parts,bike,same day,west,clear,297.0,46.96,1970-01-01 00:00:00.000000008,1970-01-01 00:00:00.000000008,no,delivered,3,1632.7206
1,250.99,xpressbees,cosmetics,ev van,express,central,cold,89.6,47.39,1970-01-01 00:00:00.000000002,1970-01-01 00:00:00.000000003,no,delivered,5,640.1700
2,250.99,shadowfax,groceries,truck,two day,east,rainy,273.5,26.89,1970-01-01 00:00:00.000000010,1970-01-01 00:00:00.000000016,no,delivered,4,1448.1700
3,250.99,dhl,electronics,ev van,same day,east,cold,269.7,12.69,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000008,no,delivered,3,1486.5700
4,250.99,dhl,clothing,van,two day,north,foggy,256.7,37.02,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000016,no,delivered,4,1394.5600


In [ ]:
df.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 15 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   delivery_id          25000 non-null  float64
 1   delivery_partner     25000 non-null  object 
 2   package_type         25000 non-null  object 
 3   vehicle_type         25000 non-null  object 
 4   delivery_mode        25000 non-null  object 
 5   region               25000 non-null  object 
 6   weather_condition    25000 non-null  object 
 7   distance_km          25000 non-null  float64
 8   package_weight_kg    25000 non-null  float64
 9   delivery_time_hours  25000 non-null  object 
 10  expected_time_hours  25000 non-null  object 
 11  delayed              25000 non-null  object 
 12  delivery_status      25000 non-null  object 
 13  delivery_rating      25000 non-null  int64  
 14  delivery_cost        25000 non-null  float64
dtypes: float64(4), int64(1), object(10)


In [ ]:
df.isnull().sum()



,0
delivery_id,0
delivery_partner,0
package_type,0
vehicle_type,0
delivery_mode,0
region,0
weather_condition,0
distance_km,0
package_weight_kg,0
delivery_time_hours,0


In [ ]:
df.describe(include = 'all')



,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,expected_time_hours,delayed,delivery_status,delivery_rating,delivery_cost
count,25000.000000,25000,25000,25000,25000,25000,25000,25000.000000,25000.000000,25000,25000,25000,25000,25000.000000,25000.000000
unique,NaN,9,9,6,4,5,6,NaN,NaN,20,9,2,3,NaN,NaN
top,NaN,xpressbees,fragile items,ev bike,two day,west,foggy,NaN,NaN,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000016,no,delivered,NaN,NaN
freq,NaN,2826,2848,4218,6302,5095,4219,NaN,NaN,3118,6302,18331,18331,NaN,NaN
mean,12500.500000,NaN,NaN,NaN,NaN,NaN,NaN,150.390436,25.145898,NaN,NaN,NaN,NaN,3.666000,864.944579
std,7212.732314,NaN,NaN,NaN,NaN,NaN,NaN,86.409745,14.368663,NaN,NaN,NaN,NaN,1.149964,435.712593
min,250.990000,NaN,NaN,NaN,NaN,NaN,NaN,3.600000,0.670000,NaN,NaN,NaN,NaN,1.000000,95.667400
25%,6250.750000,NaN,NaN,NaN,NaN,NaN,NaN,75.900000,12.680000,NaN,NaN,NaN,NaN,3.000000,490.800000
50%,12500.500000,NaN,NaN,NaN,NaN,NaN,NaN,151.000000,25.145000,NaN,NaN,NaN,NaN,4.000000,867.535000
75%,18750.250000,NaN,NaN,NaN,NaN,NaN,NaN,224.900000,37.660000,NaN,NaN,NaN,NaN,5.000000,1237.910000


In [ ]:
df.head()

,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,expected_time_hours,delayed,delivery_status,delivery_rating,delivery_cost
0,250.99,delhivery,automobile parts,bike,same day,west,clear,297.0,46.96,1970-01-01 00:00:00.000000008,1970-01-01 00:00:00.000000008,no,delivered,3,1632.7206
1,250.99,xpressbees,cosmetics,ev van,express,central,cold,89.6,47.39,1970-01-01 00:00:00.000000002,1970-01-01 00:00:00.000000003,no,delivered,5,640.1700
2,250.99,shadowfax,groceries,truck,two day,east,rainy,273.5,26.89,1970-01-01 00:00:00.000000010,1970-01-01 00:00:00.000000016,no,delivered,4,1448.1700
3,250.99,dhl,electronics,ev van,same day,east,cold,269.7,12.69,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000008,no,delivered,3,1486.5700
4,250.99,dhl,clothing,van,two day,north,foggy,256.7,37.02,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000016,no,delivered,4,1394.5600


In [ ]:
import pandas as pd
# 1. Convert the object to a proper 'Datetime' object
# Pandas will automatically see these as nanoseconds from 1970-01-01
df['delivery_time_dt'] = pd.to_datetime(df['delivery_time_hours'])
df['expected_time_dt'] = pd.to_datetime(df['expected_time_hours'])
# 2. Extract the 'nanosecond' part.
# In this specific dataset, 1 nanosecond = 1 hour of delivery time.
df['delivery_time_h'] = df['delivery_time_dt'].dt.nanosecond
df['expected_time_h'] = df['expected_time_dt'].dt.nanosecond
# 3. Validation Check
# Let's compare the 'Old' column with our 'New' clean column
print("--- Validation Check ---")
print(df[['delivery_time_hours', 'delivery_time_h']].head())

--- Validation Check ---
             delivery_time_hours  delivery_time_h
0  1970-01-01 00:00:00.000000008                8
1  1970-01-01 00:00:00.000000002                2
2  1970-01-01 00:00:00.000000010               10
3  1970-01-01 00:00:00.000000006                6
4  1970-01-01 00:00:00.000000009                9


In [ ]:
df.head()



,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,expected_time_hours,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h
0,250.99,delhivery,automobile parts,bike,same day,west,clear,297.0,46.96,1970-01-01 00:00:00.000000008,1970-01-01 00:00:00.000000008,no,delivered,3,1632.7206,1970-01-01 00:00:00.000000008,1970-01-01 00:00:00.000000008,8,8
1,250.99,xpressbees,cosmetics,ev van,express,central,cold,89.6,47.39,1970-01-01 00:00:00.000000002,1970-01-01 00:00:00.000000003,no,delivered,5,640.1700,1970-01-01 00:00:00.000000002,1970-01-01 00:00:00.000000003,2,3
2,250.99,shadowfax,groceries,truck,two day,east,rainy,273.5,26.89,1970-01-01 00:00:00.000000010,1970-01-01 00:00:00.000000016,no,delivered,4,1448.1700,1970-01-01 00:00:00.000000010,1970-01-01 00:00:00.000000016,10,16
3,250.99,dhl,electronics,ev van,same day,east,cold,269.7,12.69,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000008,no,delivered,3,1486.5700,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000008,6,8
4,250.99,dhl,clothing,van,two day,north,foggy,256.7,37.02,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000016,no,delivered,4,1394.5600,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000016,9,16


# The Delay Difference Between the delivery_time and the expected_time in Hrs

In [ ]:

# 4. Now calculate the actual Business Delay
df['delay_diff'] = df['delivery_time_h'] - df['expected_time_h']
df.head()



,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,expected_time_hours,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff
0,250.99,delhivery,automobile parts,bike,same day,west,clear,297.0,46.96,1970-01-01 00:00:00.000000008,1970-01-01 00:00:00.000000008,no,delivered,3,1632.7206,1970-01-01 00:00:00.000000008,1970-01-01 00:00:00.000000008,8,8,0
1,250.99,xpressbees,cosmetics,ev van,express,central,cold,89.6,47.39,1970-01-01 00:00:00.000000002,1970-01-01 00:00:00.000000003,no,delivered,5,640.1700,1970-01-01 00:00:00.000000002,1970-01-01 00:00:00.000000003,2,3,-1
2,250.99,shadowfax,groceries,truck,two day,east,rainy,273.5,26.89,1970-01-01 00:00:00.000000010,1970-01-01 00:00:00.000000016,no,delivered,4,1448.1700,1970-01-01 00:00:00.000000010,1970-01-01 00:00:00.000000016,10,16,-6
3,250.99,dhl,electronics,ev van,same day,east,cold,269.7,12.69,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000008,no,delivered,3,1486.5700,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000008,6,8,-2
4,250.99,dhl,clothing,van,two day,north,foggy,256.7,37.02,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000016,no,delivered,4,1394.5600,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000016,9,16,-7


In [ ]:

# See the unique categories and how many deliveries were made for each
print("--- Package Categories Count ---")
print(df['package_type'].value_counts())

print("\n--- Vehicle Fleet Count ---")
print(df['vehicle_type'].value_counts())



--- Package Categories Count ---
package_type
fragile items       2848
pharmacy            2810
documents           2805
automobile parts    2795
electronics         2792
clothing            2767
furniture           2746
cosmetics           2744
groceries           2693
Name: count, dtype: int64

--- Vehicle Fleet Count ---
vehicle_type
ev bike    4218
van        4187
scooter    4174
bike       4160
truck      4145
ev van     4116
Name: count, dtype: int64


In [ ]:

# Define what we consider "Light Items"
light_items = ['documents', 'pharmacy', 'cosmetics']

# Define "Heavy Vehicles" that consume lots of fuel
heavy_vehicles = ['truck', 'van']

# Identify trips where we used a Heavy Vehicle for a Light Item
money_leaks = df[(df['package_type'].isin(light_items)) & (df['vehicle_type'].isin(heavy_vehicles))]

# Calculate the potential wasted cost
total_wasted_trips = len(money_leaks)
total_wasted_money = money_leaks['delivery_cost'].sum()

print(f"--- Profit Leak Report ---")
print(f"Total 'High-Waste' Trips identified: {total_wasted_trips}")
print(f"Total revenue spent on inefficient fuel trips: ₹{total_wasted_money:,.2f}")

# Breakdown by Region to see where the manager is failing
print("\nWaste Breakdown by Region:")
print(money_leaks['region'].value_counts())



--- Profit Leak Report ---
Total 'High-Waste' Trips identified: 2813
Total revenue spent on inefficient fuel trips: ₹2,461,922.19

Waste Breakdown by Region:
region
south      577
central    569
north      569
west       550
east       548
Name: count, dtype: int64


In [ ]:

# Real 2026 Metrics
diesel_cost_per_km = 5.84   # Diesel Van cost per km
ev_bike_cost_per_km = 1.20  # EV Bike cost per km

# Total Distance of the 2,813 High-Waste trips
# (Assuming average trip distance from your distance_km column)
total_dist = money_leaks['distance_km'].sum()

# 1. What we ACTUALLY spent (Diesel)
actual_diesel_spend = total_dist * diesel_cost_per_km

# 2. What we COULD have spent (EV Bike)
potential_ev_spend = total_dist * ev_bike_cost_per_km

# 3. The Real-World Saving
net_savings = actual_diesel_spend - potential_ev_spend

print(f"--- Real-World Waste Analysis (North Region) ---")
print(f"Total Distance Wasted: {total_dist:,.2f} km")
print(f"Current Actual Spend (Diesel): ₹{actual_diesel_spend:,.2f}")
print(f"Potential Spend (EV Transition): ₹{potential_ev_spend:,.2f}")
print(f"NET PROFIT OPPORTUNITY: ₹{net_savings:,.2f}")



--- Real-World Waste Analysis (North Region) ---
Total Distance Wasted: 427,863.10 km
Current Actual Spend (Diesel): ₹2,498,720.50
Potential Spend (EV Transition): ₹513,435.72
NET PROFIT OPPORTUNITY: ₹1,985,284.78


In [ ]:

# 1. Calculate the 'Real' cost per kilometer for every single trip (25,000 rows)
df['real_cost_per_km'] = df['delivery_cost'] / df['distance_km']

# 2. Create a "Cost Matrix"
# This finds the average cost per km for every combination of Vehicle + Package
cost_matrix = df.groupby(['vehicle_type', 'package_type'])['real_cost_per_km'].mean().unstack()

# 3. View the Matrix
print("--- The Precision Cost Matrix (₹ per KM) ---")
display(cost_matrix)




--- The Precision Cost Matrix (₹ per KM) ---


package_type,automobile parts,clothing,cosmetics,documents,electronics,fragile items,furniture,groceries,pharmacy
vehicle_type,,,,,,,,,
bike,7.279889,6.732572,7.016989,7.405884,6.933987,7.045000,6.851446,6.822775,7.028817
ev bike,6.836486,7.094375,6.839326,7.126516,7.208338,7.027699,7.130697,7.476644,7.067066
ev van,7.372581,6.998783,7.214969,7.202233,7.209610,6.935183,6.547826,7.078283,7.060120
scooter,6.865909,7.233352,6.946315,6.951062,6.840684,7.066675,7.073361,6.877392,7.004319
truck,6.695881,7.009828,7.150463,7.096320,6.971257,7.183409,7.058269,7.089301,7.147543
van,7.355275,7.256473,6.896136,6.965865,7.107674,7.131530,6.995530,7.240634,6.536617


In [ ]:

# 1. Identify the 'High-Waste' trips again
light_items = ['documents', 'pharmacy', 'cosmetics']
heavy_vehicles = ['truck', 'van']
money_leaks = df[(df['package_type'].isin(light_items)) & (df['vehicle_type'].isin(heavy_vehicles))].copy()

# 2. Calculate what we ACTUALLY spent on these 2,813 trips
actual_spend = money_leaks['delivery_cost'].sum()

# 3. Calculate what it WOULD have cost if we used an 'EV Bike' for these EXACT trips
# We pull the specific rate for (EV Bike + Package Type) from our cost_matrix
def calculate_optimized_cost(row):
    # Get the specific rate for an EV Bike carrying that specific package type
    optimized_rate = cost_matrix.loc['ev bike', row['package_type']]
    return row['distance_km'] * optimized_rate

money_leaks['optimized_cost'] = money_leaks.apply(calculate_optimized_cost, axis=1)
optimized_total_spend = money_leaks['optimized_cost'].sum()

# 4. Final Business Result
total_savings = actual_spend - optimized_total_spend

print(f"--- 2026 Logistics Optimization Audit ---")
print(f"Total High-Waste Deliveries: {len(money_leaks)}")
print(f"Actual Money Spent (Inefficient): ₹{actual_spend:,.2f}")
print(f"Optimized Money Spent (EV Bike):   ₹{optimized_total_spend:,.2f}")
print(f"NET PROFIT RECOVERED:             ₹{total_savings:,.2f}")




--- 2026 Logistics Optimization Audit ---
Total High-Waste Deliveries: 2813
Actual Money Spent (Inefficient): ₹2,461,922.19
Optimized Money Spent (EV Bike):   ₹3,000,285.51
NET PROFIT RECOVERED:             ₹-538,363.31


In [ ]:

# Find the absolute cheapest vehicle for each package type
cheapest_vehicles = cost_matrix.idxmin()
cheapest_rates = cost_matrix.min()

def calculate_best_fit_cost(row):
    # Find the lowest possible rate for this specific package type
    best_rate = cheapest_rates[row['package_type']]
    return row['distance_km'] * best_rate

money_leaks['best_fit_cost'] = money_leaks.apply(calculate_best_fit_cost, axis=1)
best_fit_total = money_leaks['best_fit_cost'].sum()
real_savings = actual_spend - best_fit_total

print(f"--- Final Precision Optimization ---")
print(f"Actual Spend: ₹{actual_spend:,.2f}")
print(f"Best-Fit Spend: ₹{best_fit_total:,.2f}")
print(f"TRUE PROFIT RECOVERED: ₹{real_savings:,.2f}")



--- Final Precision Optimization ---
Actual Spend: ₹2,461,922.19
Best-Fit Spend: ₹2,896,805.52
TRUE PROFIT RECOVERED: ₹-434,883.33


# For Fragile Items

In [ ]:
# Filtering the dataset for the specific combination requested
specific_data = df[(df['package_type'] == 'fragile items') & (df['vehicle_type'] == 'ev bike')]

# 1. Show the total count of these specific deliveries
print(f"Total rows found for 'Fragile Items' on 'EV Bikes': {len(specific_data)}")

# 2. Display the top 10 rows to inspect the raw values
print("\n--- Raw Data Snippet (Top 10 Rows) ---")
display(specific_data.head(10))

# 3. Get the 'Data-Driven' summary for this specific pair
print("\n--- Operational Stats for the fragile items and the ev bike ---")
print(f"Average Package Weight: {specific_data['package_weight_kg'].mean():.2f} kg")
print(f"Average Delivery Cost:  ₹{specific_data['delivery_cost'].mean():,.2f}")
print(f"Average Distance:       {specific_data['distance_km'].mean():.2f} km")

Total rows found for 'Fragile Items' on 'EV Bikes': 479

--- Raw Data Snippet (Top 10 Rows) ---


,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,...,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff,real_cost_per_km
17,250.99,blue dart,fragile items,ev bike,standard,central,rainy,170.7,45.62,1970-01-01 00:00:00.000000011,...,no,delivered,5,990.3600,1970-01-01 00:00:00.000000011,1970-01-01 00:00:00.000000024,11,24,-13,5.801757
188,250.99,ekart,fragile items,ev bike,two day,north,clear,67.1,32.91,1970-01-01 00:00:00.000000002,...,no,delivered,4,434.2300,1970-01-01 00:00:00.000000002,1970-01-01 00:00:00.000000016,2,16,-14,6.471386
192,250.99,shadowfax,fragile items,ev bike,same day,north,foggy,71.0,32.56,1970-01-01 00:00:00.000000005,...,no,delivered,3,552.6800,1970-01-01 00:00:00.000000005,1970-01-01 00:00:00.000000008,5,8,-3,7.784225
243,250.99,shadowfax,fragile items,ev bike,two day,north,rainy,183.1,38.15,1970-01-01 00:00:00.000000013,...,no,delivered,4,1029.9500,1970-01-01 00:00:00.000000013,1970-01-01 00:00:00.000000016,13,16,-3,5.625068
275,276.00,fedex,fragile items,ev bike,standard,east,hot,297.1,10.39,1970-01-01 00:00:00.000000005,...,no,delivered,4,1523.1700,1970-01-01 00:00:00.000000005,1970-01-01 00:00:00.000000024,5,24,-19,5.126792
340,341.00,shadowfax,fragile items,ev bike,same day,west,clear,193.3,8.82,1970-01-01 00:00:00.000000004,...,no,delivered,4,1092.9600,1970-01-01 00:00:00.000000004,1970-01-01 00:00:00.000000008,4,8,-4,5.654216
356,357.00,blue dart,fragile items,ev bike,express,north,stormy,293.3,47.56,1970-01-01 00:00:00.000000015,...,yes,delayed,3,1632.7206,1970-01-01 00:00:00.000000015,1970-01-01 00:00:00.000000007,15,7,8,5.566726
493,494.00,shadowfax,fragile items,ev bike,two day,south,hot,197.1,39.53,1970-01-01 00:00:00.000000007,...,no,delivered,4,1104.0900,1970-01-01 00:00:00.000000007,1970-01-01 00:00:00.000000016,7,16,-9,5.601674
575,576.00,dhl,fragile items,ev bike,same day,north,rainy,232.2,1.33,1970-01-01 00:00:00.000000009,...,yes,delayed,3,1264.9900,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000008,9,8,1,5.447847
588,589.00,amazon logistics,fragile items,ev bike,express,south,foggy,268.6,42.34,1970-01-01 00:00:00.000000011,...,yes,delayed,3,1520.0200,1970-01-01 00:00:00.000000011,1970-01-01 00:00:00.000000007,11,7,4,5.659047



--- Operational Stats for the fragile items and the ev bike ---
Average Package Weight: 25.35 kg
Average Delivery Cost:  ₹864.86
Average Distance:       150.04 km


In [ ]:

# Filtering the dataset for the specific combination requested
specific_data = df[(df['package_type'] == 'fragile items') & (df['vehicle_type'] == 'bike')]

# 1. Show the total count of these specific deliveries
print(f"Total rows found for 'Fragile Items' on 'Bikes': {len(specific_data)}")

# 2. Display the top 10 rows to inspect the raw values
print("\n--- Raw Data Snippet (Top 10 Rows) ---")
display(specific_data.head(10))

# 3. Get the 'Data-Driven' summary for this specific pair
print("\n--- Operational Stats for the fragile items and the bike ---")
print(f"Average Package Weight: {specific_data['package_weight_kg'].mean():.2f} kg")
print(f"Average Delivery Cost:  ₹{specific_data['delivery_cost'].mean():,.2f}")
print(f"Average Distance:       {specific_data['distance_km'].mean():.2f} km")




Total rows found for 'Fragile Items' on 'Bikes': 450

--- Raw Data Snippet (Top 10 Rows) ---


,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,...,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff,real_cost_per_km
93,250.99,dhl,fragile items,bike,two day,north,hot,83.3,34.71,1970-01-01 00:00:00.000000004,...,no,delivered,4,520.6300,1970-01-01 00:00:00.000000004,1970-01-01 00:00:00.000000016,4,16,-12,6.250060
183,250.99,xpressbees,fragile items,bike,standard,south,stormy,228.6,5.64,1970-01-01 00:00:00.000000009,...,no,delivered,4,1159.9200,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000024,9,24,-15,5.074016
186,250.99,blue dart,fragile items,bike,two day,north,rainy,80.7,0.67,1970-01-01 00:00:00.000000007,...,no,delivered,4,404.7600,1970-01-01 00:00:00.000000007,1970-01-01 00:00:00.000000016,7,16,-9,5.015613
228,250.99,fedex,fragile items,bike,express,east,hot,225.4,22.30,1970-01-01 00:00:00.000000005,...,no,delivered,5,1243.9000,1970-01-01 00:00:00.000000005,1970-01-01 00:00:00.000000006,5,6,-1,5.518634
411,412.00,ekart,fragile items,bike,standard,east,cold,297.1,46.75,1970-01-01 00:00:00.000000009,...,no,delivered,3,1632.7206,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000024,9,24,-15,5.495525
430,431.00,delhivery,fragile items,bike,same day,west,foggy,224.2,13.03,1970-01-01 00:00:00.000000010,...,yes,failed,1,1260.0900,1970-01-01 00:00:00.000000010,1970-01-01 00:00:00.000000008,10,8,2,5.620384
434,435.00,dhl,fragile items,bike,express,south,foggy,94.2,32.09,1970-01-01 00:00:00.000000005,...,yes,failed,1,617.2700,1970-01-01 00:00:00.000000005,1970-01-01 00:00:00.000000003,5,3,2,6.552760
440,441.00,ecom express,fragile items,bike,same day,south,stormy,149.7,38.51,1970-01-01 00:00:00.000000011,...,yes,delayed,2,964.0300,1970-01-01 00:00:00.000000011,1970-01-01 00:00:00.000000008,11,8,3,6.439746
446,447.00,ekart,fragile items,bike,two day,east,stormy,49.9,40.86,1970-01-01 00:00:00.000000007,...,no,delivered,5,372.0800,1970-01-01 00:00:00.000000007,1970-01-01 00:00:00.000000016,7,16,-9,7.456513
488,489.00,xpressbees,fragile items,bike,standard,west,foggy,218.2,5.67,1970-01-01 00:00:00.000000009,...,no,delivered,4,1108.0100,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000024,9,24,-15,5.077956



--- Operational Stats for the fragile items and the bike ---
Average Package Weight: 26.16 kg
Average Delivery Cost:  ₹850.89
Average Distance:       147.05 km


In [ ]:

# Filtering the dataset for the specific combination requested
specific_data = df[(df['package_type'] == 'fragile items') & (df['vehicle_type'] == 'ev van')]

# 1. Show the total count of these specific deliveries
print(f"Total rows found for 'Fragile Items' on 'EV Vans': {len(specific_data)}")

# 2. Display the top 10 rows to inspect the raw values
print("\n--- Raw Data Snippet (Top 10 Rows) ---")
display(specific_data.head(10))

# 3. Get the 'Data-Driven' summary for this specific pair
print("\n--- Operational Stats for the fragile items and the ev van ---")
print(f"Average Package Weight: {specific_data['package_weight_kg'].mean():.2f} kg")
print(f"Average Delivery Cost:  ₹{specific_data['delivery_cost'].mean():,.2f}")
print(f"Average Distance:       {specific_data['distance_km'].mean():.2f} km")




Total rows found for 'Fragile Items' on 'EV Vans': 456

--- Raw Data Snippet (Top 10 Rows) ---


,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,...,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff,real_cost_per_km
234,250.99,fedex,fragile items,ev van,standard,central,foggy,267.0,26.49,1970-01-01 00:00:00.000000011,...,no,delivered,3,1414.47,1970-01-01 00:00:00.000000011,1970-01-01 00:00:00.000000024,11,24,-13,5.297640
264,265.00,ecom express,fragile items,ev van,express,east,clear,133.4,24.33,1970-01-01 00:00:00.000000004,...,yes,failed,2,789.99,1970-01-01 00:00:00.000000004,1970-01-01 00:00:00.000000004,4,4,0,5.921964
481,482.00,amazon logistics,fragile items,ev van,two day,south,cold,71.9,31.14,1970-01-01 00:00:00.000000004,...,no,delivered,4,452.92,1970-01-01 00:00:00.000000004,1970-01-01 00:00:00.000000016,4,16,-12,6.299305
490,491.00,ecom express,fragile items,ev van,express,south,cold,99.6,2.60,1970-01-01 00:00:00.000000003,...,no,delivered,4,555.80,1970-01-01 00:00:00.000000003,1970-01-01 00:00:00.000000003,3,3,0,5.580321
552,553.00,amazon logistics,fragile items,ev van,express,south,stormy,264.0,14.26,1970-01-01 00:00:00.000000012,...,yes,delayed,3,1412.78,1970-01-01 00:00:00.000000012,1970-01-01 00:00:00.000000007,12,7,5,5.351439
626,627.00,dhl,fragile items,ev van,standard,central,clear,24.7,47.66,1970-01-01 00:00:00.000000001,...,no,delivered,5,266.48,1970-01-01 00:00:00.000000001,1970-01-01 00:00:00.000000024,1,24,-23,10.788664
695,696.00,ekart,fragile items,ev van,same day,east,cold,55.2,26.94,1970-01-01 00:00:00.000000003,...,no,delivered,3,456.82,1970-01-01 00:00:00.000000003,1970-01-01 00:00:00.000000008,3,8,-5,8.275725
716,717.00,fedex,fragile items,ev van,same day,east,foggy,166.2,4.50,1970-01-01 00:00:00.000000007,...,no,delivered,4,944.50,1970-01-01 00:00:00.000000007,1970-01-01 00:00:00.000000008,7,8,-1,5.682912
916,917.00,fedex,fragile items,ev van,two day,north,hot,121.8,19.36,1970-01-01 00:00:00.000000004,...,no,delivered,5,667.08,1970-01-01 00:00:00.000000004,1970-01-01 00:00:00.000000016,4,16,-12,5.476847
919,920.00,shadowfax,fragile items,ev van,standard,central,stormy,139.9,8.36,1970-01-01 00:00:00.000000006,...,no,delivered,4,724.58,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000024,6,24,-18,5.179271



--- Operational Stats for the fragile items and the ev van ---
Average Package Weight: 25.23 kg
Average Delivery Cost:  ₹859.28
Average Distance:       149.82 km


In [ ]:

# Filtering the dataset for the specific combination requested
specific_data = df[(df['package_type'] == 'fragile items') & (df['vehicle_type'] == 'van')]

# 1. Show the total count of these specific deliveries
print(f"Total rows found for 'Fragile Items' on 'Vans': {len(specific_data)}")

# 2. Display the top 10 rows to inspect the raw values
print("\n--- Raw Data Snippet (Top 10 Rows) ---")
display(specific_data.head(10))

# 3. Get the 'Data-Driven' summary for this specific pair
print("\n--- Operational Stats for the fragile items and the van ---")
print(f"Average Package Weight: {specific_data['package_weight_kg'].mean():.2f} kg")
print(f"Average Delivery Cost:  ₹{specific_data['delivery_cost'].mean():,.2f}")
print(f"Average Distance:       {specific_data['distance_km'].mean():.2f} km")



Total rows found for 'Fragile Items' on 'Vans': 485

--- Raw Data Snippet (Top 10 Rows) ---


,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,...,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff,real_cost_per_km
7,250.99,xpressbees,fragile items,van,same day,north,cold,114.6,42.63,1970-01-01 00:00:00.000000004,...,no,delivered,3,800.89,1970-01-01 00:00:00.000000004,1970-01-01 00:00:00.000000008,4,8,-4,6.988569
178,250.99,ecom express,fragile items,van,same day,south,foggy,225.0,7.32,1970-01-01 00:00:00.000000011,...,yes,delayed,3,1246.96,1970-01-01 00:00:00.000000011,1970-01-01 00:00:00.000000008,11,8,3,5.542044
185,250.99,fedex,fragile items,van,express,central,foggy,250.7,29.97,1970-01-01 00:00:00.000000010,...,yes,failed,2,1393.41,1970-01-01 00:00:00.000000010,1970-01-01 00:00:00.000000007,10,7,3,5.558077
219,250.99,ecom express,fragile items,van,express,central,clear,207.8,40.12,1970-01-01 00:00:00.000000006,...,yes,delayed,2,1209.36,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000006,6,6,0,5.819827
254,255.00,shadowfax,fragile items,van,two day,east,hot,224.0,31.45,1970-01-01 00:00:00.000000004,...,no,delivered,3,1214.35,1970-01-01 00:00:00.000000004,1970-01-01 00:00:00.000000016,4,16,-12,5.421205
362,363.00,ekart,fragile items,van,two day,south,foggy,247.4,26.71,1970-01-01 00:00:00.000000007,...,no,delivered,5,1317.13,1970-01-01 00:00:00.000000007,1970-01-01 00:00:00.000000016,7,16,-9,5.323888
453,454.00,delhivery,fragile items,van,standard,south,cold,158.0,14.34,1970-01-01 00:00:00.000000003,...,no,delivered,5,833.02,1970-01-01 00:00:00.000000003,1970-01-01 00:00:00.000000024,3,24,-21,5.272278
482,483.00,ekart,fragile items,van,express,central,foggy,201.4,39.58,1970-01-01 00:00:00.000000010,...,yes,delayed,3,1175.74,1970-01-01 00:00:00.000000010,1970-01-01 00:00:00.000000006,10,6,4,5.837835
528,529.00,ecom express,fragile items,van,same day,west,stormy,181.0,3.79,1970-01-01 00:00:00.000000011,...,yes,delayed,3,1016.37,1970-01-01 00:00:00.000000011,1970-01-01 00:00:00.000000008,11,8,3,5.615304
723,724.00,blue dart,fragile items,van,two day,east,clear,146.9,40.17,1970-01-01 00:00:00.000000003,...,no,delivered,4,855.01,1970-01-01 00:00:00.000000003,1970-01-01 00:00:00.000000016,3,16,-13,5.820354



--- Operational Stats for the fragile items and the van ---
Average Package Weight: 26.43 kg
Average Delivery Cost:  ₹846.94
Average Distance:       146.23 km


In [ ]:
# Filtering the dataset for the specific combination requested
specific_data = df[(df['package_type'] == 'fragile items') & (df['vehicle_type'] == 'truck')]

# 1. Show the total count of these specific deliveries
print(f"Total rows found for 'Fragile Items' on 'Truck': {len(specific_data)}")

# 2. Display the top 10 rows to inspect the raw values
print("\n--- Raw Data Snippet (Top 10 Rows) ---")
display(specific_data.head(10))

# 3. Get the 'Data-Driven' summary for this specific pair
print("\n--- Operational Stats for the fragile items and the Truck ---")
print(f"Average Package Weight: {specific_data['package_weight_kg'].mean():.2f} kg")
print(f"Average Delivery Cost:  ₹{specific_data['delivery_cost'].mean():,.2f}")
print(f"Average Distance:       {specific_data['distance_km'].mean():.2f} km")





Total rows found for 'Fragile Items' on 'Truck': 480

--- Raw Data Snippet (Top 10 Rows) ---


,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,...,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff,real_cost_per_km
42,250.99,xpressbees,fragile items,truck,standard,west,stormy,258.5,8.37,1970-01-01 00:00:00.000000007,...,no,delivered,5,1317.61,1970-01-01 00:00:00.000000007,1970-01-01 00:00:00.000000024,7,24,-17,5.097137
60,250.99,ecom express,fragile items,truck,same day,central,foggy,9.3,47.62,1970-01-01 00:00:00.000000003,...,no,delivered,5,289.36,1970-01-01 00:00:00.000000003,1970-01-01 00:00:00.000000008,3,8,-5,31.113978
124,250.99,amazon logistics,fragile items,truck,two day,south,hot,59.5,7.79,1970-01-01 00:00:00.000000002,...,no,delivered,3,320.87,1970-01-01 00:00:00.000000002,1970-01-01 00:00:00.000000016,2,16,-14,5.392773
134,250.99,amazon logistics,fragile items,truck,same day,west,hot,127.8,20.03,1970-01-01 00:00:00.000000002,...,no,delivered,3,799.09,1970-01-01 00:00:00.000000002,1970-01-01 00:00:00.000000008,2,8,-6,6.252660
198,250.99,dhl,fragile items,truck,express,central,foggy,198.5,20.10,1970-01-01 00:00:00.000000004,...,no,delivered,5,1102.80,1970-01-01 00:00:00.000000004,1970-01-01 00:00:00.000000005,4,5,-1,5.555668
206,250.99,xpressbees,fragile items,truck,express,east,rainy,24.5,0.67,1970-01-01 00:00:00.000000005,...,yes,delayed,2,174.12,1970-01-01 00:00:00.000000005,1970-01-01 00:00:00.000000002,5,2,3,7.106939
242,250.99,delhivery,fragile items,truck,same day,north,hot,34.6,47.42,1970-01-01 00:00:00.000000003,...,no,delivered,4,415.26,1970-01-01 00:00:00.000000003,1970-01-01 00:00:00.000000008,3,8,-5,12.001734
256,257.00,xpressbees,fragile items,truck,two day,north,cold,145.2,46.17,1970-01-01 00:00:00.000000005,...,no,delivered,3,864.51,1970-01-01 00:00:00.000000005,1970-01-01 00:00:00.000000016,5,16,-11,5.953926
266,267.00,fedex,fragile items,truck,two day,south,cold,22.1,11.47,1970-01-01 00:00:00.000000001,...,no,delivered,5,144.91,1970-01-01 00:00:00.000000001,1970-01-01 00:00:00.000000016,1,16,-15,6.557014
391,392.00,blue dart,fragile items,truck,express,north,rainy,188.6,21.04,1970-01-01 00:00:00.000000011,...,yes,delayed,2,1056.12,1970-01-01 00:00:00.000000011,1970-01-01 00:00:00.000000005,11,5,6,5.599788



--- Operational Stats for the fragile items and the Truck ---
Average Package Weight: 25.23 kg
Average Delivery Cost:  ₹841.77
Average Distance:       145.76 km


In [ ]:

# Filtering the dataset for the specific combination requested
specific_data = df[(df['package_type'] == 'fragile items') & (df['vehicle_type'] == 'scooter')]

# 1. Show the total count of these specific deliveries
print(f"Total rows found for 'Fragile Items' on 'Scooter': {len(specific_data)}")

# 2. Display the top 10 rows to inspect the raw values
print("\n--- Raw Data Snippet (Top 10 Rows) ---")
display(specific_data.head(10))

# 3. Get the 'Data-Driven' summary for this specific pair
print("\n--- Operational Stats for the fragile items and the Scooter ---")
print(f"Average Package Weight: {specific_data['package_weight_kg'].mean():.2f} kg")
print(f"Average Delivery Cost:  ₹{specific_data['delivery_cost'].mean():,.2f}")
print(f"Average Distance:       {specific_data['distance_km'].mean():.2f} km")




Total rows found for 'Fragile Items' on 'Scooter': 498

--- Raw Data Snippet (Top 10 Rows) ---


,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,...,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff,real_cost_per_km
77,250.99,ekart,fragile items,scooter,standard,east,rainy,197.1,28.23,1970-01-01 00:00:00.000000013,...,no,delivered,4,1070.19,1970-01-01 00:00:00.000000013,1970-01-01 00:00:00.000000024,13,24,-11,5.429680
194,250.99,ecom express,fragile items,scooter,two day,north,cold,157.0,43.03,1970-01-01 00:00:00.000000003,...,no,delivered,4,914.09,1970-01-01 00:00:00.000000003,1970-01-01 00:00:00.000000016,3,16,-13,5.822229
231,250.99,amazon logistics,fragile items,scooter,standard,central,clear,262.7,28.70,1970-01-01 00:00:00.000000005,...,no,delivered,5,1399.60,1970-01-01 00:00:00.000000005,1970-01-01 00:00:00.000000024,5,24,-19,5.327750
424,425.00,dhl,fragile items,scooter,two day,north,foggy,193.9,30.19,1970-01-01 00:00:00.000000005,...,no,delivered,4,1060.07,1970-01-01 00:00:00.000000005,1970-01-01 00:00:00.000000016,5,16,-11,5.467096
469,470.00,ecom express,fragile items,scooter,same day,north,foggy,257.7,39.95,1970-01-01 00:00:00.000000013,...,yes,failed,1,1508.35,1970-01-01 00:00:00.000000013,1970-01-01 00:00:00.000000008,13,8,5,5.853124
480,481.00,ekart,fragile items,scooter,standard,east,hot,160.0,40.30,1970-01-01 00:00:00.000000006,...,no,delivered,4,920.90,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000024,6,24,-18,5.755625
618,619.00,delhivery,fragile items,scooter,two day,north,cold,59.5,3.06,1970-01-01 00:00:00.000000003,...,no,delivered,5,306.68,1970-01-01 00:00:00.000000003,1970-01-01 00:00:00.000000016,3,16,-13,5.154286
628,629.00,fedex,fragile items,scooter,two day,west,rainy,45.7,42.70,1970-01-01 00:00:00.000000005,...,no,delivered,3,356.60,1970-01-01 00:00:00.000000005,1970-01-01 00:00:00.000000016,5,16,-11,7.803063
635,636.00,xpressbees,fragile items,scooter,standard,central,clear,152.3,35.01,1970-01-01 00:00:00.000000005,...,no,delivered,5,866.53,1970-01-01 00:00:00.000000005,1970-01-01 00:00:00.000000024,5,24,-19,5.689626
653,654.00,amazon logistics,fragile items,scooter,same day,east,rainy,187.3,39.79,1970-01-01 00:00:00.000000010,...,yes,failed,2,1155.87,1970-01-01 00:00:00.000000010,1970-01-01 00:00:00.000000008,10,8,2,6.171223



--- Operational Stats for the fragile items and the Scooter ---
Average Package Weight: 23.61 kg
Average Delivery Cost:  ₹822.76
Average Distance:       142.85 km


# For Documents

In [ ]:

# Fetching: Documents on a standard Bike
docs_bike = df[(df['package_type'] == 'documents') & (df['vehicle_type'] == 'bike')]
print(f"Total 'Documents' on 'Bikes': {len(docs_bike)}")
print(f"Avg Weight: {docs_bike['package_weight_kg'].mean():.2f} kg | Avg Cost: ₹{docs_bike['delivery_cost'].mean():.2f}")
display(docs_bike.head(5))



Total 'Documents' on 'Bikes': 456
Avg Weight: 25.16 kg | Avg Cost: ₹822.48


,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,...,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff,real_cost_per_km
16,250.99,fedex,documents,bike,standard,west,stormy,79.0,0.73,1970-01-01 00:00:00.000000010,...,no,delivered,5,397.19,1970-01-01 00:00:00.000000010,1970-01-01 00:00:00.000000024,10,24,-14,5.027722
35,250.99,delhivery,documents,bike,same day,east,clear,37.6,22.81,1970-01-01 00:00:00.000000003,...,no,delivered,4,356.43,1970-01-01 00:00:00.000000003,1970-01-01 00:00:00.000000008,3,8,-5,9.479521
38,250.99,ekart,documents,bike,standard,east,rainy,62.2,48.58,1970-01-01 00:00:00.000000007,...,no,delivered,3,456.74,1970-01-01 00:00:00.000000007,1970-01-01 00:00:00.000000024,7,24,-17,7.343087
69,250.99,fedex,documents,bike,same day,central,cold,217.3,3.03,1970-01-01 00:00:00.000000006,...,no,delivered,5,1195.59,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000008,6,8,-2,5.502025
81,250.99,dhl,documents,bike,same day,east,rainy,60.4,22.93,1970-01-01 00:00:00.000000006,...,no,delivered,5,470.79,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000008,6,8,-2,7.794536


In [ ]:

# Fetching: Documents on an EV Bike
docs_ev_bike = df[(df['package_type'] == 'documents') & (df['vehicle_type'] == 'ev bike')]
print(f"Total 'Documents' on 'EV Bikes': {len(docs_ev_bike)}")
print(f"Avg Weight: {docs_ev_bike['package_weight_kg'].mean():.2f} kg | Avg Cost: ₹{docs_ev_bike['delivery_cost'].mean():.2f}")
display(docs_ev_bike.head(5))



Total 'Documents' on 'EV Bikes': 460
Avg Weight: 25.79 kg | Avg Cost: ₹872.87


,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,...,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff,real_cost_per_km
5,250.99,amazon logistics,documents,ev bike,express,west,rainy,48.4,33.15,1970-01-01 00:00:00.000000004,...,yes,delayed,3,391.45,1970-01-01 00:00:00.000000004,1970-01-01 00:00:00.000000002,4,2,2,8.087810
78,250.99,shadowfax,documents,ev bike,same day,west,foggy,94.8,10.02,1970-01-01 00:00:00.000000002,...,no,delivered,4,604.06,1970-01-01 00:00:00.000000002,1970-01-01 00:00:00.000000008,2,8,-6,6.371941
115,250.99,shadowfax,documents,ev bike,same day,south,clear,65.3,45.09,1970-01-01 00:00:00.000000004,...,no,delivered,5,561.77,1970-01-01 00:00:00.000000004,1970-01-01 00:00:00.000000008,4,8,-4,8.602910
162,250.99,dhl,documents,ev bike,two day,east,hot,230.8,30.01,1970-01-01 00:00:00.000000006,...,no,delivered,4,1244.03,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000016,6,16,-10,5.390078
171,250.99,delhivery,documents,ev bike,two day,south,hot,3.6,49.52,1970-01-01 00:00:00.000000001,...,no,delivered,5,163.39,1970-01-01 00:00:00.000000001,1970-01-01 00:00:00.000000016,1,16,-15,45.386111


In [ ]:

# Fetching: Documents on a Scooter
docs_scooter = df[(df['package_type'] == 'documents') & (df['vehicle_type'] == 'scooter')]
print(f"Total 'Documents' on 'Scooters': {len(docs_scooter)}")
print(f"Avg Weight: {docs_scooter['package_weight_kg'].mean():.2f} kg | Avg Cost: ₹{docs_scooter['delivery_cost'].mean():.2f}")
display(docs_scooter.head(5))




Total 'Documents' on 'Scooters': 445
Avg Weight: 24.38 kg | Avg Cost: ₹859.50


,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,...,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff,real_cost_per_km
47,250.99,ecom express,documents,scooter,two day,north,rainy,187.9,14.96,1970-01-01 00:00:00.000000011,...,no,delivered,4,984.38,1970-01-01 00:00:00.000000011,1970-01-01 00:00:00.000000016,11,16,-5,5.238850
166,250.99,ekart,documents,scooter,express,west,rainy,182.1,2.26,1970-01-01 00:00:00.000000011,...,yes,delayed,2,967.28,1970-01-01 00:00:00.000000011,1970-01-01 00:00:00.000000005,11,5,6,5.311807
237,250.99,dhl,documents,scooter,same day,central,rainy,124.2,1.02,1970-01-01 00:00:00.000000007,...,no,delivered,5,724.06,1970-01-01 00:00:00.000000007,1970-01-01 00:00:00.000000008,7,8,-1,5.829791
357,358.00,ecom express,documents,scooter,express,central,foggy,228.0,20.54,1970-01-01 00:00:00.000000008,...,yes,delayed,3,1251.62,1970-01-01 00:00:00.000000008,1970-01-01 00:00:00.000000006,8,6,2,5.489561
363,364.00,ecom express,documents,scooter,two day,north,stormy,165.5,12.13,1970-01-01 00:00:00.000000013,...,no,delivered,4,863.89,1970-01-01 00:00:00.000000013,1970-01-01 00:00:00.000000016,13,16,-3,5.219879


In [ ]:

# Fetching: Documents on an EV Van
docs_ev_van = df[(df['package_type'] == 'documents') & (df['vehicle_type'] == 'ev van')]
print(f"Total 'Documents' on 'EV Vans': {len(docs_ev_van)}")
print(f"Avg Weight: {docs_ev_van['package_weight_kg'].mean():.2f} kg | Avg Cost: ₹{docs_ev_van['delivery_cost'].mean():.2f}")
display(docs_ev_van.head(5))




Total 'Documents' on 'EV Vans': 489
Avg Weight: 23.86 kg | Avg Cost: ₹862.91


,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,...,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff,real_cost_per_km
25,250.99,shadowfax,documents,ev van,two day,north,stormy,115.4,0.67,1970-01-01 00:00:00.000000006,...,no,delivered,4,578.59,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000016,6,16,-10,5.013778
48,250.99,xpressbees,documents,ev van,express,west,hot,78.3,49.22,1970-01-01 00:00:00.000000004,...,yes,failed,1,589.16,1970-01-01 00:00:00.000000004,1970-01-01 00:00:00.000000003,4,3,1,7.524393
72,250.99,fedex,documents,ev van,express,north,rainy,78.5,27.87,1970-01-01 00:00:00.000000008,...,yes,delayed,3,526.11,1970-01-01 00:00:00.000000008,1970-01-01 00:00:00.000000003,8,3,5,6.702038
121,250.99,xpressbees,documents,ev van,standard,north,cold,102.0,18.67,1970-01-01 00:00:00.000000004,...,no,delivered,4,566.01,1970-01-01 00:00:00.000000004,1970-01-01 00:00:00.000000024,4,24,-20,5.549118
157,250.99,shadowfax,documents,ev van,same day,south,stormy,165.4,43.20,1970-01-01 00:00:00.000000011,...,yes,delayed,2,1056.60,1970-01-01 00:00:00.000000011,1970-01-01 00:00:00.000000008,11,8,3,6.388150


In [ ]:

# Fetching: Documents on a standard Diesel Van
docs_van = df[(df['package_type'] == 'documents') & (df['vehicle_type'] == 'van')]
print(f"Total 'Documents' on 'Vans': {len(docs_van)}")
print(f"Avg Weight: {docs_van['package_weight_kg'].mean():.2f} kg | Avg Cost: ₹{docs_van['delivery_cost'].mean():.2f}")
display(docs_van.head(5))




Total 'Documents' on 'Vans': 485
Avg Weight: 25.32 kg | Avg Cost: ₹825.48


,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,...,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff,real_cost_per_km
18,250.99,amazon logistics,documents,van,same day,north,rainy,90.0,33.70,1970-01-01 00:00:00.000000004,...,no,delivered,3,651.10,1970-01-01 00:00:00.000000004,1970-01-01 00:00:00.000000008,4,8,-4,7.234444
73,250.99,blue dart,documents,van,express,north,hot,138.3,47.87,1970-01-01 00:00:00.000000008,...,yes,delayed,3,885.11,1970-01-01 00:00:00.000000008,1970-01-01 00:00:00.000000004,8,4,4,6.399928
88,250.99,xpressbees,documents,van,standard,east,rainy,111.0,36.68,1970-01-01 00:00:00.000000009,...,no,delivered,4,665.04,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000024,9,24,-15,5.991351
114,250.99,fedex,documents,van,express,central,stormy,259.3,27.01,1970-01-01 00:00:00.000000008,...,yes,failed,1,1427.53,1970-01-01 00:00:00.000000008,1970-01-01 00:00:00.000000007,8,7,1,5.505322
217,250.99,delhivery,documents,van,standard,central,foggy,137.4,27.87,1970-01-01 00:00:00.000000007,...,no,delivered,3,770.61,1970-01-01 00:00:00.000000007,1970-01-01 00:00:00.000000024,7,24,-17,5.608515


In [ ]:

# Fetching: Documents on a Heavy Truck
docs_truck = df[(df['package_type'] == 'documents') & (df['vehicle_type'] == 'truck')]
print(f"Total 'Documents' on 'Trucks': {len(docs_truck)}")
print(f"Avg Weight: {docs_truck['package_weight_kg'].mean():.2f} kg | Avg Cost: ₹{docs_truck['delivery_cost'].mean():.2f}")
display(docs_truck.head(5))




Total 'Documents' on 'Trucks': 470
Avg Weight: 24.97 kg | Avg Cost: ₹862.17


,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,...,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff,real_cost_per_km
147,250.99,ekart,documents,truck,same day,south,foggy,147.7,46.57,1970-01-01 00:00:00.000000006,...,no,delivered,4,978.21,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000008,6,8,-2,6.622952
184,250.99,fedex,documents,truck,express,north,rainy,169.6,41.44,1970-01-01 00:00:00.000000009,...,yes,delayed,3,1022.32,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000005,9,5,4,6.027830
200,250.99,dhl,documents,truck,same day,north,clear,265.3,31.80,1970-01-01 00:00:00.000000005,...,no,delivered,5,1521.90,1970-01-01 00:00:00.000000005,1970-01-01 00:00:00.000000008,5,8,-3,5.736525
248,250.99,amazon logistics,documents,truck,same day,east,foggy,123.5,29.37,1970-01-01 00:00:00.000000008,...,yes,delayed,2,805.61,1970-01-01 00:00:00.000000008,1970-01-01 00:00:00.000000008,8,8,0,6.523158
257,258.00,delhivery,documents,truck,two day,central,stormy,204.9,41.15,1970-01-01 00:00:00.000000009,...,no,delivered,4,1147.95,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000016,9,16,-7,5.602489


# Summary :
# Vehicle Type  Total Rows  Avg. Weight   Avg. Delivery Cost
# Bike              456      25.16 kg          ₹822.48
# EV Bike           460      25.79 kg          ₹872.87
# Scooter           445      24.38 kg          ₹859.50
# EV Van            489      23.86 kg          ₹862.91
# Van               485      25.32 kg          ₹825.48
# Truck             470      24.97 kg          ₹862.17

In [ ]:
df.head()

,delivery_id,delivery_partner,package_type,vehicle_type,delivery_mode,region,weather_condition,distance_km,package_weight_kg,delivery_time_hours,...,delayed,delivery_status,delivery_rating,delivery_cost,delivery_time_dt,expected_time_dt,delivery_time_h,expected_time_h,delay_diff,real_cost_per_km
0,250.99,delhivery,automobile parts,bike,same day,west,clear,297.0,46.96,1970-01-01 00:00:00.000000008,...,no,delivered,3,1632.7206,1970-01-01 00:00:00.000000008,1970-01-01 00:00:00.000000008,8,8,0,5.497376
1,250.99,xpressbees,cosmetics,ev van,express,central,cold,89.6,47.39,1970-01-01 00:00:00.000000002,...,no,delivered,5,640.1700,1970-01-01 00:00:00.000000002,1970-01-01 00:00:00.000000003,2,3,-1,7.144754
2,250.99,shadowfax,groceries,truck,two day,east,rainy,273.5,26.89,1970-01-01 00:00:00.000000010,...,no,delivered,4,1448.1700,1970-01-01 00:00:00.000000010,1970-01-01 00:00:00.000000016,10,16,-6,5.294954
3,250.99,dhl,electronics,ev van,same day,east,cold,269.7,12.69,1970-01-01 00:00:00.000000006,...,no,delivered,3,1486.5700,1970-01-01 00:00:00.000000006,1970-01-01 00:00:00.000000008,6,8,-2,5.511939
4,250.99,dhl,clothing,van,two day,north,foggy,256.7,37.02,1970-01-01 00:00:00.000000009,...,no,delivered,4,1394.5600,1970-01-01 00:00:00.000000009,1970-01-01 00:00:00.000000016,9,16,-7,5.432645


In [ ]:
from google.colab import files

# 1. Save the DataFrame to a CSV with a name that matches your actual project
# 'India_Logistics_Cleaned_Audit' is much more professional
df.to_csv('India_Logistics_Cleaned_Audit.csv', index=False)

# 2. Trigger the download to your computer
files.download('India_Logistics_Cleaned_Audit.csv')

print("Your cleaned Logistics dataset is now being downloaded!")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Your cleaned Logistics dataset is now being downloaded!
